### DISCUSSION TITLE
How to Fix Misplacement of Position and Rotation in Curved IfcBeam Geometry?

### DISCUSSION THREAD

**Discussion by ibrahimdh2** [IfcOpenShell/IfcOpenShell#5578] [Q&A]
I am extracting the geometry of a whole Ifc file and using that to generate a 3D model in Unity. It is working well so far; however, there are some curved IfcBeams that get misplaced (seems like their position and rotation gets messed up).

(I am not using IfcConvert because I need to categorize parts of the model under certain groups like meshes belonging to a certain IfcBuildingStorey) and mesh that are visible during floor plan)

**Comment by Moult** [comment]
Can you provide a sample file and show which beams are misplaced? i.e. where they should be vs where they are.

**Reply by ibrahimdh2** [reply]
> Can you provide a sample file and show which beams are misplaced? i.e. where they should be vs where they are.

How it looks in Open Ifc Viewer: 
<img width="354" alt="ifcopenshell" src="https://github.com/user-attachments/assets/68362de7-7c50-4015-8abe-b4106b15bc76">

How it looks in Unity (constructed from geometry data - everything else works fine but only these curved beams)
<img width="564" alt="Unity" src="https://github.com/user-attachments/assets/10a0e9b0-ada2-4175-aedf-119cfddcddbf">

Their rotation and position changed

**Reply by Moult** [reply]
You'll need to provide a file, otherwise it's hard to say exactly what the issue is unfortunately.

**Reply by ibrahimdh2** [reply]
> You'll need to provide a file, otherwise it's hard to say exactly what the issue is unfortunately.

Here's the file: https://we.tl/t-ps9Oym6Qyi

**Reply by Moult** [reply]
Everything looks OK here. What version are you using? How are you extracting data?

FYI you can use Python's serialisation bindings to do more complex conversions.

![2024-10-16-100743_864x715_scrot](https://github.com/user-attachments/assets/6fa76031-6388-4fa7-a248-863e50c073de)

**Reply by ibrahimdh2** [reply]
> Everything looks OK here. What version are you using? How are you extracting data?
> 
> FYI you can use Python's serialisation bindings to do more complex conversions.
> 
> ![2024-10-16-100743_864x715_scrot](https://private-user-images.githubusercontent.com/88302/376828055-6fa76031-6388-4fa7-a248-863e50c073de.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwMzg3NTMsIm5iZiI6MTcyOTAzODQ1MywicGF0aCI6Ii84ODMwMi8zNzY4MjgwNTUtNmZhNzYwMzEtNjM4OC00ZmE3LWEyNDgtODYzZTUwYzA3M2RlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE2VDAwMjczM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiMmE4ZjE4MDIwODg4MzAwZDE4NzM4MzM1YzliMzY4MzI3YjcyZTExMWRlYTJjNTZhYTczNzliYTg5ZDk4YWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QCK0moUeE8UBAEFovJ7pUbzBSdqspOYqDNHlV6SMIiE)

I am using ifcopenshell 0.8.0, and this is what I am doing recursively getting each element's geometry and storing it in json. 
Basically, getting each IfcBuildingStorey and storing geometry of all elements in a story in a JSON. Then using the vertices and faces data to generate mesh in Unity.

```
 def get_geometry(element):

        geometry_data = {}

        shape = ifcopenshell.geom.create_shape(IFCProcessor.settings, element)
        geometry_data["vertices"] = shape.geometry.verts
        geometry_data["faces"] = shape.geometry.faces
        return geometry_data
        
```

**Reply by ibrahimdh2** [reply]
Also whenever I try to use serialization as per the [[documentation)](https://docs.ifcopenshell.org/ifcopenshell-python/geometry_processing.html) I get these errors - 


    raise AttributeError("'Settings' object has no attribute '%s'" % k)
AttributeError: 'Settings' object has no attribute 'STRICT_TOLERANCE'

AttributeError: 'Settings' object has no attribute 'INCLUDE_CURVES'

AttributeError: 'Settings' object has no attribute 'INCLUDE_CURVES

**Reply by Moult** [reply]
Using create shape is a very slow / inefficient way to create shapes, and you're missing the multiplication by the element matrix which explains why your arches have wonky locations / rotations.

I recommend to use the geometry iterator: https://docs.ifcopenshell.org/ifcopenshell-python/geometry_processing.html#geometry-iterator and be sure to use the `shape.transformation.matrix` matrix.

**Reply by ibrahimdh2** [reply]
> Using create shape is a very slow / inefficient way to create shapes, and you're missing the multiplication by the element matrix which explains why your arches have wonky locations / rotations.
> 
> I recommend to use the geometry iterator: https://docs.ifcopenshell.org/ifcopenshell-python/geometry_processing.html#geometry-iterator and be sure to use the `shape.transformation.matrix` matrix.

I forgot to mention but I already am storing position and rotation as well:



 ```
def get_position(element):
        matrix = ifcopenshell.util.placement.get_local_placement(element.ObjectPlacement)
        position = matrix[:, 3][:3]  # Extract X, Y, Z position
       
        return [float(coord) / 1000.0 for coord in position
        
        def get_rotation(element):
        matrix = ifcopenshell.util.placement.get_local_placement(element.ObjectPlacement)
        rotation_matrix = matrix[:3, :3]

        def rotation_matrix_to_euler_angles(rot):
            sy = np.sqrt(rot[0, 0] * rot[0, 0] + rot[1, 0] * rot[1, 0])
            singular = sy < 1e-6  # Check for gimbal lock

            if not singular:
                x = np.arctan2(rot[2, 1], rot[2, 2])
                y = np.arctan2(-rot[2, 0], sy)
                z = np.arctan2(rot[1, 0], rot[0, 0])
            else:
                x = np.arctan2(-rot[1, 2], rot[1, 1])
                y = np.arctan2(-rot[2, 0], sy)
                z = 0

            return np.degrees(x), np.degrees(y), np.degrees(z)
        
```

**Reply by Moult** [reply]
There's probably an error in your rotation code. It's worth mentioning that it's better to just get the matrix from the iterator, it'll be faster. Also the hardcoded division by 1000 is probably not a good idea, use `util.unit.calculate_unit_scale`.
